src/openfermion/utils/_trotter_exp_to_qgates.py

```diff
@@ -0,0 +1,329 @@
+        term_ordering = sorted(list(hamiltonian.terms.keys()))
+    if len(term_ordering) == 0:
+        raise TypeError("term_ordering must None or non-empty list.")
+
+    ret_val = []
+    # First order trotter
+    if trotter_order == 1:
+        for step in range(trotter_number):
+            for op in term_ordering:
+                ret_val.append(QubitOperator(op,
+                                             hamiltonian.terms[op]
+                                             * float(k_exp)
+                                             / float(trotter_number)
+                                             ))
+    # Second order trotter
+    elif trotter_order == 2:
+        if len(term_ordering) < 2:
+            raise ValueError("Not enough terms in the Hamiltonian to do " +
+                             "second order trotterization")
+
+        for op in term_ordering[:-1]:
+            ret_val.append(QubitOperator(op,
+                                         (hamiltonian.terms[op]
+                                          * k_exp
+                                          / (2.0 * float(trotter_number)))
+                                         ))
+
+        ret_val.append(QubitOperator(term_ordering[-1],
+                       hamiltonian.terms[term_ordering[-1]] * float(k_exp)
+                       / float(trotter_number)))
+
+        for op in reversed(term_ordering[:-1]):
+            ret_val.append(QubitOperator(op, hamiltonian.terms[op]
+                                         * float(k_exp)
+                                         / (2.0 * float(trotter_number))))
+        ret_val_copy = []
+        ret_val_copy.extend(ret_val)
+        for step in range(trotter_number - 1):
+            ret_val.extend(ret_val_copy)
+
+    # Third order trotter
+    elif trotter_order == 3:
+        if len(term_ordering) < 2:
+            raise ValueError("Not enough terms in the Hamiltonian to do " +
+                             "third order trotterization")
+        # Make sure original hamiltonian is not modified
+        ham = hamiltonian / float(trotter_number)
+        ham *= float(k_exp)
+        atemp = QubitOperator(term_ordering[0], ham.terms[term_ordering[0]])
+        btemp = []
+        for term in term_ordering[1:]:
+            btemp.append(QubitOperator(term, ham.terms[term]))
+        ret_val = third_order_trotter_helper(atemp, btemp)
+        ret_val_copy = []
+        ret_val_copy.extend(ret_val)
+        for step in range(trotter_number - 1):
+            ret_val.extend(ret_val_copy)
+
+    # Return list, each member a single Pauli-string QubitOperator
+    return ret_val
+
+
+def exp_sgl_pauli_string_to_qasm(output_stream,
+                                 qubit_operator):
+    """Exponentiate a single-Pauli-string QubitOperator
+    and print it in QASM format to a supplied output stream
+
+    Args:
+        output_stream (file/string/IO stream): QASM output stream
```

Comment:
I would keep the output parameters last if you need them.  It would be nice if all these routines just returned their value though, the user can always just dump it to a file if they want, but getting it back into a string using stringio is kind of a pain.

Reply:
So you think scrap the filestream entirely? Just return a string?

Reply:
I think scrap the filestream and construct the functions as generators that yield each instruction

Reply:
Totally agree after reading more on python generators - implementing now